localstack/ext/java/src/main/java/cloud/localstack/docker/LocalstackDockerTestRunner.java

```diff
@@ -0,0 +1,198 @@
+package cloud.localstack.docker;
+
+import java.lang.annotation.Annotation;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+import org.junit.runner.notification.RunNotifier;
+import org.junit.runners.BlockJUnit4ClassRunner;
+import org.junit.runners.model.InitializationError;
+
+import cloud.localstack.docker.annotation.IHostNameResolver;
+import cloud.localstack.docker.annotation.LocalstackDockerProperties;
+
+/**
+ * JUnit test runner that automatically pulls and runs the latest localstack docker image
+ * and then terminates when tests are complete.
+ *
+ * Having docker installed is a prerequisite for this test runner to execute.  If docker
+ * is not installed in one of the default locations (C:\program files\docker\docker\resources\bin\, usr/local/bin or usr/bin)
+ * then use the DOCKER_LOCATION environment variable to specify the location.
+ *
+ * Since ports are dynamically allocated, the external port needs to be resolved based on the default localstack port.
+ *
+ * The hostname defaults to localhost, but in some environments that is not sufficient, so the HostName can be specified
+ * by using the LocalstackDockerProperties annotation with an IHostNameResolver.
+ *
+ * @author Alan Bevier
+ */
+public class LocalstackDockerTestRunner extends BlockJUnit4ClassRunner {
+
+    private static final Logger LOG = Logger.getLogger(LocalstackDockerTestRunner.class.getName());
+
+    //These are the default internal ports in the localstack docker image, in order to 
+    //use them, they must be resolved to an external docker port via Container.getExternalPortFor()
+    public static final int API_GATEWAY_INTERNAL_PORT = 4567;
```

Comment:
I'd very much prefer not to define these constants in multiple places. Any chance we can download the following config file, cache it in a temporary local file, and parse/extract the default ports from there?
https://raw.githubusercontent.com/localstack/localstack-python-client/master/localstack_client/config.py

Another option would be to re-use the `getEndpoint` method from [`LocalstackTestRunner`](https://github.com/localstack/localstack/blob/master/localstack/ext/java/src/main/java/cloud/localstack/LocalstackTestRunner.java#L187-L199). Note that this method additionally considers the `SERVICES` configuration variable (to customize the service ports), which means that we may have to adjust the initialization of `Container.ports`, and pass in the `SERVICES` environment variable in `createLocalstackContainer`.

Reply:
Since localstack is running in a container, I'm not sure that customizing the internal ports is necessary or desirable.  Docker is already exposing random external points that map to these internal ports so no conflict is possible.  

Additionally altering the internal ports would cause issue with my assumption on line 45 of Container.java.  If internal ports were altered, the ports that are exposed by the docker command would also need to be changed.

I could possibly cat out the config.py file from the docker container and parse the settings from there?  That would allow us to avoid hardcoding these constants, but there would still be an issue with exposing the correct ports on the container.